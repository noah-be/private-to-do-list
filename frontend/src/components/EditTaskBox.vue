<template>
  <div class="edit-task-modal">
    <div class="edit-task-modal-content">
      <span class="close" @click="$emit('close')">&times;</span>
      <h2>Edit Task</h2>
      <form>
        <div>
          <label for="edit-task-title">Title:</label>
          <input type="text" id="edit-task-title" v-model="taskCopy.title" />
        </div>
        <div>
          <label for="edit-task-description">Description:</label>
          <textarea
            id="edit-task-description"
            v-model="taskCopy.description"
          ></textarea>
        </div>
        <div>
          <label for="edit-task-due-date">Due Date:</label>
          <input
            type="date"
            id="edit-task-due-date"
            v-model="taskCopy.dueDate"
          />
        </div>
        <div>
          <label for="edit-task-priority">Priority:</label>
          <select id="edit-task-priority" v-model="taskCopy.priority">
            <option value="High">High</option>
            <option value="Medium">Medium</option>
            <option value="Low">Low</option>
          </select>
        </div>
        <button type="button" @click="saveChanges">Save Changes</button>
      </form>
    </div>
  </div>
</template>

<script>
export default {
  props: {
    task: {
      type: Object,
      required: true,
    },
  },
  data() {
    return {
      taskCopy: { ...this.task },
    };
  },
  methods: {
    saveChanges() {
      this.$emit("update-task", this.taskCopy);
      this.$emit("close");
    },
  },
};
</script>
