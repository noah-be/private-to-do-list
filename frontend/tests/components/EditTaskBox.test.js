import { describe, it, expect, vi } from "vitest";
import { mount } from "@vue/test-utils";
import EditTaskBox from "@/components/EditTaskBox.vue";

describe("EditTaskBox.vue", () => {
  const mockTask = {
    _id: "1",
    title: "Test Task",
    description: "Test Description",
    dueDate: "2023-12-31",
    priority: "High",
  };

  it("matches the snapshot", () => {
    const wrapper = mount(EditTaskBox, {
      props: {
        task: mockTask,
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it("emits save-task with updated task on saveChanges", async () => {
    const updatedTask = { ...mockTask, title: "Updated Task" };

    global.fetch = vi.fn(() =>
      Promise.resolve({
        ok: true,
        json: () => Promise.resolve(updatedTask),
      }),
    );

    const wrapper = mount(EditTaskBox, {
      props: { task: mockTask },
    });

    await wrapper.find("#edit-task-title").setValue("Updated Task");
    await wrapper.find("#save-edit-task-btn").trigger("click");

    expect(global.fetch).toHaveBeenCalledWith(
      `/api/task/${mockTask._id}`,
      expect.objectContaining({
        method: "PATCH",
        headers: {
          "Content-Type": "application/json",
          Authorization: `Bearer ${localStorage.getItem("token")}`,
        },
        body: JSON.stringify(updatedTask),
      }),
    );

    expect(wrapper.emitted("save-task")).toBeTruthy();
    expect(wrapper.emitted("save-task")[0][0].title).toBe("Updated Task");

    global.fetch.mockRestore();
  });

  it("emits delete-task with task ID on deleteTask", async () => {
    const wrapper = mount(EditTaskBox, {
      props: { task: mockTask },
    });

    global.fetch = vi.fn(() =>
      Promise.resolve({
        ok: true,
      }),
    );

    await wrapper.find("#delete-task-btn").trigger("click");

    expect(global.fetch).toHaveBeenCalledWith(
      `/api/task/${mockTask._id}`,
      expect.anything(),
    );
    expect(wrapper.emitted("delete-task")).toBeTruthy();
    expect(wrapper.emitted("delete-task")[0][0]).toBe(mockTask._id);
  });

  it("emits close on clicking close button", async () => {
    const wrapper = mount(EditTaskBox, {
      props: { task: mockTask },
    });

    await wrapper.find(".close").trigger("click");
    expect(wrapper.emitted("close")).toBeTruthy();
  });

  it("updates taskCopy when form fields are changed", async () => {
    const wrapper = mount(EditTaskBox, {
      props: { task: mockTask },
    });

    await wrapper.find("#edit-task-title").setValue("New Title");
    await wrapper.find("#edit-task-description").setValue("New Description");
    await wrapper.find("#edit-task-due-date").setValue("2024-01-01");
    await wrapper.find("#edit-task-priority").setValue("Medium");

    expect(wrapper.vm.taskCopy.title).toBe("New Title");
    expect(wrapper.vm.taskCopy.description).toBe("New Description");
    expect(wrapper.vm.taskCopy.dueDate).toBe("2024-01-01");
    expect(wrapper.vm.taskCopy.priority).toBe("Medium");
  });
});
