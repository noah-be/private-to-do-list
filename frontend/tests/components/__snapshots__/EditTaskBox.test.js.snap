// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EditTaskBox.vue > matches the snapshot 1`] = `
"<v-dialog model-value="false" max-width="500">
  <v-card>
    <v-card-title><span class="text-h6">Edit Task</span></v-card-title>
    <v-card-text>
      <v-form>
        <v-text-field modelvalue="Test Task" label="Title" outlined="" dense=""></v-text-field>
        <v-textarea modelvalue="Test Description" label="Description" outlined="" dense=""></v-textarea>
        <v-text-field modelvalue="2023-12-31" label="Due Date" type="date" outlined="" dense=""></v-text-field>
        <v-select modelvalue="High" items="High,Medium,Low" label="Priority" outlined="" dense=""></v-select>
      </v-form>
    </v-card-text>
    <v-divider></v-divider>
    <v-card-actions>
      <v-btn color="error">Delete</v-btn>
      <v-spacer></v-spacer>
      <v-btn color="success">Save Changes</v-btn>
    </v-card-actions>
  </v-card>
</v-dialog>"
`;
